src/state_machine.rs

```diff
@@ -0,0 +1,1062 @@
+                             $crate::state_machine::MapMutator::new(
+                                 stringify!($name),
+                                 $crate::sugar::NamedArguments(names, values),
+                                 |$statevar: &mut $statety,
+                                  $crate::sugar::NamedArguments(
+                                      _, proptest_helper!(@_WRAPPAT ($($parm)*)))
+                                 | {
+                                     let ret: &mut $childty = $body;
+                                     Ok(ret)
+                                 },
+                                 mutation))))));
+
+        prop_sm_helper! {
+            @_MUNCH($ops, $invariants, $statety) $($rest)*
+        }
+    };
+
+    (@_MUNCH($ops:ident, $invariants:ident, $statety:ty)
+     $(#[$attr:ident = $attrv:expr])*
+     delegate ($child_machine:expr)
+     to $name:ident (&mut $statevar:ident, $($arg:tt)*)
+                     -> &mut $childty:ty $body:block
+     $($rest:tt)*) => {
+        let names = proptest_helper!(@_EXT _STR ($($arg)*));
+
+        #[allow(unused_mut)]
+        let mut attrs = $crate::state_machine::helper::OpAttrs::DEFAULT;
+        $(attrs.$attr = $attrv;)*
+
+        $ops.push(
+            (attrs.weight,
+             $crate::strategy::Strategy::boxed(
+                 $crate::strategy::Strategy::prop_map(
+                     ($child_machine.mutation_strategy(),
+                      proptest_helper!(@_EXT _STRAT ($($arg)*))),
+                     move |(mutation, values)|
+                         $crate::state_machine::Mutator::boxed(
+                             $crate::state_machine::MapMutator::new(
+                                 stringify!($name),
+                                 $crate::sugar::NamedArguments(names, values),
+                                 |$statevar: &mut $statety,
+                                  $crate::sugar::NamedArguments(
+                                      _, proptest_helper!(@_EXT _PAT ($($arg)*)))
+                                 | {
+                                     let ret: &mut $childty = $body;
+                                     Ok(ret)
+                                 },
+                                 mutation))))));
+
+        prop_sm_helper! {
+            @_MUNCH($ops, $invariants, $statety) $($rest)*
+        }
+    };
+}
+
+/// A state-machine-based test.
+///
+/// `StateMachine` values are simultaneously a description of mutations and
+/// invariants of that state, as well as a test that the state under test
+/// upholds its expected behaviours.
+///
+/// Instances of this struct are usually constructed by the
+/// `proptest_state_machine!` macro. Please see the documentation of that macro
+/// for examples of typical usage.
+#[must_use = "StateMachine does nothing unless used"]
+pub struct StateMachine<S : fmt::Debug> {
+    mutation_strategy: BoxedStrategy<Box<dyn Mutator<S>>>,
```

Comment:
+ `BoxedStrategy<?A>` requires `?A: 'static`
+ `?A = Box<dyn 'static + Mutator<S>>`
+ This requires `S: 'static`

Proof:

```rust
use std::fmt;

trait Mutator<S> {}

impl<'a, S> Mutator<S> for &'a str {}

pub struct StateMachine<S: fmt::Debug> {
    ms: Box<dyn Mutator<S>>,
}

fn main() {
    let string: String = "foobar".into();
    let borrow: &str = &string;
    let ms: Box<dyn Mutator<&str>> = Box::new(borrow);
    // Remove the line below and it is OK!
    StateMachine { ms };
}
```

Results in:

```rust
   Compiling playground v0.0.1 (file:///playground)
error[E0597]: `string` does not live long enough
  --> src/main.rs:13:25
   |
13 |     let borrow: &str = &string;
   |                         ^^^^^^ borrowed value does not live long enough
...
17 | }
   | - borrowed value only lives until here
   |
   = note: borrowed value must be valid for the static lifetime...
```

If we change to the following it will work fine; but that might cause problems elsewhere:

```rust
pub struct StateMachine<'a, S: fmt::Debug> {
    ms: Box<dyn Mutator<S> + 'a>,
}